NGF: Update data plane configuration doc (#529)
nginx/documentation: content/ngf/how-to/data-plane-configuration.md

---

## Configuring the GatewayClass NginxProxy on install

> By default, an `NginxProxy` resource is created in the same namespace where NGINX Gateway Fabric is installed, attached to the GatewayClass. You can set configuration options in the `nginx` Helm value section, and the resource will be created and attached using the set values. You can also [manually create and attach](#manually-creating-nginxProxies) specific `NginxProxy` resources to target different Gateways.

Reviewer: Just want to confirm that this subsection link works properly?

Author: Yes, the one we had before did not. So updated it

> By default, the `NginxProxy` resource is created in the same namespace where NGINX Gateway Fabric is installed. You can set configuration options in the `nginx.` Helm values, and the resource will be created and attached using the set values, when NGINX Gateway Fabric is installed using Helm. You can also manually create and attach specific `NginxProxy` resources to target different data planes [manually creating NginxProxies](#manually-creating-nginxProxies).

Reviewer: ```suggestion
By default, an `NginxProxy` resource is created in the same namespace where NGINX Gateway Fabric is installed, attached to the GatewayClass. You can set configuration options in the `nginx` Helm value section, and the resource will be created and attached using the set values. You can also [manually create and attach](#manually-creating-nginxProxies) specific `NginxProxy` resources to target different Gateways.
```

EOF
```

---

> ### Attaching NginxProxy to GatewayClass

Reviewer: I don't think we need this section since we create one by default now that's attached to the GatewayClass.

Author: but i think this information is helpful if they need to create one manually for gateways

Reviewer: Yes, I'm just talking about the GatewayClass section

If everything is valid, the `ResolvedRefs` condition should be `True`. Otherwise, you will see an `InvalidParameters` condition in the status.

> {{< note >}} The `NginxProxy` resource must reside in the same namespace as the Gateway it is attached to. {{< /note >}}

Reviewer: I'd move this note up to be under the section where we attach to the Gateway.

Author: this is under the section `Attaching NginxProxy to Gateway` , are you seeing something different in your diff? or maybe I am misunderstanding?

Reviewer: I just mean right after we edit the Gateway resource, before we describe it.

---

## Configure the data plane log level

> You can use the `NginxProxy` resource at the GatewayClass level to dynamically configure the log level for all data plane instances.

Reviewer: This change isn't necessary. We don't need to talk about "globally" changing it. It can be changed at either global or per-Gateway.

EOF
```

> Other log levels supported are debug, notice, error, crit, alert, emerg.

Reviewer: Why remove links to the autogenerated API? If fields ever change, we're going to now have to update this doc too, instead of just the API reference.

Author: I moved the link to the bottom of the doc. maybe be I can link the section here. 

We had too many similar sections about seeing the api doc so it made sense to consolidate into one

Reviewer: I personally don't have an issue with linking multiple times, because I feel like a user may be looking at a specific section and want to know the list of fields that are available. They may not go to the bottom of the doc.

EOF
```

{{< note >}} When sending curl requests to a server expecting proxy information, use the flag `--haproxy-protocol` to avoid broken header errors. {{< /note >}}

---

## Configure infrastructure-related settings

> You can configure deployment and service settings for all data plane instances by editing the `NginxProxy` resource at the Gateway or GatewayClass level. These settings can also be specified under `nginx.` in the Helm values file. You can edit things such as replicas, pod scheduling options, container resource limits, extra volume mounts, service types and load balancer settings. 

Reviewer: ```suggestion
You can configure deployment and service settings for all data plane instances by editing the `NginxProxy` resource at the Gateway or GatewayClass level. These settings can also be specified under the `nginx` section in the Helm values file. You can edit things such as replicas, pod scheduling options, container resource limits, extra volume mounts, service types and load balancer settings. 
```

> You can configure deployment and service settings for all data plane instances by editing the `NginxProxy` resource at the Gateway or GatewayClass level. These settings can also be specified under the `nginx.` section in the Helm values file. You can edit things such as replicas, pod scheduling options, container resource limits, extra volume mounts, service types and load balancer settings. 

Reviewer: ```suggestion
You can configure deployment and service settings for all data plane instances by editing the `NginxProxy` resource at the Gateway or GatewayClass level. These settings can also be specified under the `nginx` section in the Helm values file. You can edit things such as replicas, pod scheduling options, container resource limits, extra volume mounts, service types and load balancer settings. 
```

You can configure deployment and service settings for all data plane instances by editing the `NginxProxy` resource at the Gateway or GatewayClass level. These settings can also be specified under `nginx.` in the Helm values file. You can edit things such as replicas, pod scheduling options, container resource limits, extra volume mounts, service types and load balancer settings. 

The following command creates an `NginxProxy` resource with 2 replicas, sets `container.resources.requests` to 100m CPU and 128Mi memory, configures a 90 second `pod.terminationGracePeriodSeconds`, and sets the service type to `LoadBalancer` with IP `192.87.9.1` and AWS NLB annotation.

```yaml
kubectl apply -f - <<EOF
apiVersion: gateway.nginx.org/v1alpha2
kind: NginxProxy
metadata:
  name: ngf-proxy-config-test
spec:
  kubernetes:
    deployment:
      container:
        resources:
          requests:
            cpu: 100m
            memory: 128Mi
      pod:
        terminationGracePeriodSeconds: 90
      replicas: 2
    service:
      annotations:
        service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
      type: LoadBalancer
      loadBalancerIP: "192.87.9.1"
EOF
```

> To view the full list of supported log levels, see the `NginxProxy spec` in the [API reference]({{< ref "/ngf/reference/api.md" >}}).

Reviewer: should be configuration options, not log levels

EOF
```

{{< note >}} When sending curl requests to a server expecting proxy information, use the flag `--haproxy-protocol` to avoid broken header errors. {{< /note >}}

---

## Configure infrastructure-related settings

> You can configure deployment and service settings for all data plane instances by editing the `NginxProxy` resource at the GatewayClass level. These settings can also be specified under `nginx.` in the Helm values file. 

Reviewer: I would enhance this to say that you can also set these fields per-Gateway, because that's going to be an important way to do it if a user has multiple Gateways.

Users can specify these settings for NGINX data plane deployments:

>   - _replicas_ specifies the number of data plane pod replicas..

Reviewer: Same note as above, we can maybe mention an example list like "You can edit things such as...", and mention the most important things, but the exhaustive list should just live in the API doc so we don't have to keep updating this document.

And maybe provide an example NginxProxy that sets some of those fields (like replicas, service loadBalancerIP, etc)

Users can specify these settings for NGINX data plane service in the `service` config:

  - _type_ specifies the service type for the NGINX data plane. Allowed values are ClusterIP, NodePort, or LoadBalancer.
  - _externalTrafficPolicy_ sets how external traffic is handled. `Local` preserves the client’s source IP.
  - _annotations_ adds custom annotations to the NGINX data plane service.
  - Certain fields are only applicable when _service.type_ is set to `LoadBalancer`: _loadBalancerIP_ to assign a static IP, _loadBalancerClass_ to define the load balancer implementation, and _loadBalancerSourceRanges_ to restrict access to specific IP ranges (CIDRs).

  The below fields can only be specified when the `service.type` is `LoadBalancer`:
  - _loadBalancerIP_ specifies the static IP address of the load balancer.
  - _loadBalancerClass_ specifies the class of the load balancer implementation this service belongs to.
  - _loadBalancerSourceRanges_ specifies IP ranges (CIDR) that are allowed to access the load balancer.

---

## See also

For a full list of configuration options that can be set, see the `NginxProxy spec` in the [API reference]({{< ref "/ngf/reference/api.md" >}}).

> ---

Reviewer 2: ```suggestion
```

Kind of odd to add a horizontal line at the end of the page.
